tree/dataframe/inc/ROOT/RDFAction.hxx

```diff
@@ -0,0 +1,153 @@
+// Author: Enrico Guiraud, Danilo Piparo CERN  09/2018
+
+/*************************************************************************
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_RACTION
+#define ROOT_RACTION
+
+#include "ROOT/GraphNode.hxx"
+#include "ROOT/RActionBase.hxx"
+#include "ROOT/RDFNodesUtils.hxx" // InitRDFValues
+#include "ROOT/RDFUtils.hxx" // ColumnNames_t
+#include "ROOT/RDFColumnValue.hxx"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace ROOT {
+namespace Internal {
+namespace RDF {
+
+namespace RDFDetail = ROOT::Detail::RDF;
+namespace RDFGraphDrawing = ROOT::Internal::RDF::GraphDrawing;
+
+// fwd decl for RAction, RFilter
+namespace GraphDrawing {
+std::shared_ptr<GraphNode>
+CreateDefineNode(const std::string &columnName, const RDFDetail::RCustomColumnBase *columnPtr);
+} // namespace GraphDrawing
+
+// fwd decl for RAction
+namespace GraphDrawing {
+bool CheckIfDefaultOrDSColumn(const std::string &name, const std::shared_ptr<RDFDetail::RCustomColumnBase> &column);
+} // ns GraphDrawing
+
+template <typename Helper, typename PrevDataFrame, typename ColumnTypes_t = typename Helper::ColumnTypes_t>
+class RAction final : public RActionBase {
+   using TypeInd_t = std::make_index_sequence<ColumnTypes_t::list_size>;
+
+   Helper fHelper;
+   const std::shared_ptr<PrevDataFrame> fPrevDataPtr;
+   PrevDataFrame &fPrevData;
+   std::vector<RDFValueTuple_t<ColumnTypes_t>> fValues;
+
+public:
+   RAction(Helper &&h, const ColumnNames_t &bl, std::shared_ptr<PrevDataFrame> pd,
+           const RBookedCustomColumns &customColumns)
+      : RActionBase(pd->GetLoopManagerUnchecked(), pd->GetLoopManagerUnchecked()->GetNSlots(), bl, customColumns),
+        fHelper(std::forward<Helper>(h)), fPrevDataPtr(std::move(pd)), fPrevData(*fPrevDataPtr), fValues(fNSlots) {}
+
+   RAction(const RAction &) = delete;
+   RAction &operator=(const RAction &) = delete;
```

Comment:
I wonder if `RAction` isn't too generic and may lead to conflicts later along with other such generic names.

Reply:
all our classes except a select few are in either ROOT::Internal::RDF or ROOT::Detail::RDF